lib/instrumentation/modules/mongodb.js

```diff
@@ -8,18 +8,40 @@ const HOSTNAME_PORT_RE = /^([^:]+):(\d+)$/
 
 module.exports = (mongodb, agent, { version, enabled }) => {
   if (!enabled) return mongodb
-  if (!semver.satisfies(version, '>=3.3')) {
+  if (!semver.satisfies(version, '>=3.3 < 5.0')) {
     agent.logger.debug('mongodb version %s not instrumented (mongodb <3.3 is instrumented via mongodb-core)', version)
     return mongodb
   }
 
-  const listener = mongodb.instrument()
   const activeSpans = new Map()
+  if (mongodb.instrument) {
+    const listener = mongodb.instrument()
 
-  listener.on('started', onStart)
-  listener.on('succeeded', onEnd)
-  listener.on('failed', onEnd)
-
+    listener.on('started', onStart)
+    listener.on('succeeded', onEnd)
+    listener.on('failed', onEnd)
+  } else if (mongodb.MongoClient) {
+    // mongo 4.0+ removed the instrument() method in favor of
+    // listeners on the instantiated client objects.
+    class MongoClient extends mongodb.MongoClient {
+      constructor () {
+        super(...arguments)
+        this.on('commandStarted', onStart)
+        this.on('commandSucceeded', onEnd)
+        this.on('commandFailed', onEnd)
+      }
+    }
+    Object.defineProperty(
+      mongodb,
+      'MongoClient',
+      {
+        enumerable: true,
+        get: function () {
+          return MongoClient
+        }
+      }
+    )
+  }
```

Comment:
Was there a reason to add this via defineProperty rather than something like:

```js
mongodb.MongoClient = MongoClient
// or
return Object.assign(mongodb, { MongoClient: MongoClient })
```

Reply:
Good question.  That was my first instinct as well.  However, this was the result

> TypeError: Cannot set property MongoClient of `#<Object>` which has only a getter

The mongo library/driver uses `defineProperty` to set those 

```
    #File: node_modules/mongodb/lib/index.js
    /* ... */
    Object.defineProperty(exports, "MongoClient", { enumerable: true, get: function () { return mongo_client_1.MongoClient; } });
    /* ... */
```

which means (or appears to mean?) that we need to redefine the property.

(Also, it's worth noting that it looks like MongoDB has moved to a pure typescript implementation, and the commonjs libraries are the results of compilation.)

Reply:
Ah, fair. This sounds part of the path I went through with `@aws-sdk/client-s3` described at https://github.com/elastic/apm-agent-nodejs/blob/35a5a56b9b58e079740d15adedf2f7fad80ea1f6/lib/instrumentation/modules/%40aws-sdk/client-s3.js#L178-L221



> (Also, it's worth noting that it looks like MongoDB has moved to a pure typescript implementation, and the commonjs libraries are the results of compilation.)

Ah. Hence the `Object.defineProperty(exports, "MongoClient",  ...` for properties. I'm slightly curious why defineProperty worked for you on `mongodb` but did not for me for client-s3. Hrm.

Reply:
> I'm slightly curious why defineProperty worked for you on mongodb but did not for me for client-s3. Hrm.

An excellent nerd snipe.  It looks like it's this line that's the difference.

> Gist of compiled lib/index.js from mongodb package
> https://gist.github.com/astorm/bfe2ed7ac34445046867000176693c8e#file-mongodb-lib-index-js-L3

It seems like most/every exported property is first assigned a value of `void 0`.  This appears to have the side effect of making the property [configurable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty).  Without this line I also see the `TypeError: Cannot redefine property:` errors. 

So Re _worked for you on mongodb but did not for me for client-s3_ -- luck and compiler flags.

Reply:
WAT. Wow, excellent snipe response.